### PR TITLE
chore: add circleci based deploy

### PR DESCRIPTION
**Pull request by olizilla**
ci is dead. long live ci.

License: MIT
Signed-off-by: Oli Evans <oli@tableflip.io>

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            hash=$(ipfs-cluster-ctl \
+              --host /dnsaddr/cluster.ipfs.io \
+              --basic-auth $CLUSTER_USER:$CLUSTER_PASSWORD \
+              add --rmin 3 --rmax 3 --name "$DOMAIN" \
+              --recursive /tmp/workspace/$BUILD_DIR | tail -n1 | cut -d " " -f 2 )
+
+            preview_url=https://ipfs.io/ipfs/$hash
+
+            # Preload ipfs hash on the gateway
+            curl --silent --output /dev/null --show-error $preview_url
```
Comment (lidel) [inline .circleci/config.yml]: Was the intention to preload just the HTML of main page?
If not, entire thing could be preloaded asynchronously in best-effort fashion without breaking the build:

```bash
nohup curl --max-time 900 -s "https://ipfs.io/api/v0/refs?r=true&arg=${$preview_url}" >/dev/null 2>&1 &
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      - persist_to_workspace:
+          root: .
+          paths:
+            - build
+
+  deploy:
+    docker:
+      - image: olizilla/ipfs-dns-deploy
+        environment:
+          DOMAIN: explore.ipld.io
```
Comment (fsdiogo) [inline .circleci/config.yml]: `webui.ipfs.io`?

Comment (olizilla) [reply .circleci/config.yml]: derp

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      - attach_workspace:
+          at: /tmp/workspace
+      - run:
+          name: Pin website, post notification for PRs or update DNS on master
+          command: |
+            # Upload build dir to cluster.ipfs.io
+            hash=$(ipfs-cluster-ctl \
+              --host /dnsaddr/cluster.ipfs.io \
+              --basic-auth $CLUSTER_USER:$CLUSTER_PASSWORD \
+              add --rmin 3 --rmax 3 --name "$DOMAIN" \
```
Comment (hsanjuan) [inline .circleci/config.yml]: Do not set --rmin/--rmax, that way cluster's defaults will be used.. Also, if by chance you have the [short]-commit ID available, you could add it to the `--name` tag so there is more info.

File: .circleci/config.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+              --recursive /tmp/workspace/$BUILD_DIR | tail -n1 | cut -d " " -f 2 )
```
Comment (hsanjuan) [inline .circleci/config.yml]: You want to use the `--quietest` or `--Q` option which only outputs the final hash, rather than `tail ...`

**Comment by olizilla** [comment]
Thanks @hsanjuan that's good to know. Will update.

**Pull request merged**